### PR TITLE
Gmm train with weights

### PR DESCRIPTION
**Pull request by charlotte-avery**
Expand the repeat in batches based on the weights before training:
- Weights = number of time each sample occurs in the dataset
- Before training GMM, expand the repeat samples in the batch based on the sample weights
- Shuffle the expanded batch to prevent consecutive repeated samples which may impact GMM convergence

Abstract the `prepare_data_for_model` code. 

Abstract FaradayVAE module into separate script to prevent circular imports.
- `src/opensynth/models/faraday/model.py` depends on `_lightning.py` scripts which themselves take FaradayVAE as input. Before FaradayVAE was defined in `src/opensynth/models/faraday/model.py` creating a circular import. 
- I have moved FaradayVAE to `vae_models.py` to prevent circular import.

### CODE REVIEW

**Comment by Ianlmgoddard** [comment]
Do we want to target `develop` for this PR or `feature-gmm-lightning`?

**Comment by charlotte-avery** [comment]
> Do we want to target `develop` for this PR or `feature-gmm-lightning`?

`feature-gmm-lightning` because I am updating scripts in the `gaussian_mixture` directory which is not in dev

File: src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from opensynth.models.faraday.losses import _expand_samples
+from opensynth.models.faraday.model import FaradayVAE
+
+
+def prepare_data_for_model(
+    vae_module: FaradayVAE, data: torch.Tensor
+) -> torch.Tensor:
+    """Prepare data for the GMM by encoding it with the VAE.
+     If sample weights are used, then expand the repeat based on the weights
+      and shuffle the expanded dataset.
```
Comment (Ianlmgoddard) [inline src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py:12]:
```suggestion
    """Prepare data for the GMM by encoding it with the VAE.
     If sample weights are used, then expand the repeat based on the weights
     and shuffle the expanded dataset.
```

Just fixing indenting

File: src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import torch
+
+from opensynth.models.faraday.losses import _expand_samples
+from opensynth.models.faraday.model import FaradayVAE
+
+
+def prepare_data_for_model(
```
Comment (Ianlmgoddard) [inline src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py:7]: We should probably add some tests for this function

File: src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py (hunk 1 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py]: Updated to suggested changes.

File: src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    features = data["features"]
+    vae_input = vae_module.reshape_data(kwh, features)
+    vae_output = vae_module.encode(vae_input)
+    model_input = vae_module.reshape_data(vae_output, features)
+    if "weights" in data:
+        weights = data["weights"]
+        model_input = _expand_samples(model_input, weights)
+        model_input = model_input[
+            torch.randperm(model_input.size()[0])
+        ]  # Shuffle tensor
```
[the remark this answers is not in the document]
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py]: Done!

File: src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    features = data["features"]
+    vae_input = vae_module.reshape_data(kwh, features)
+    vae_output = vae_module.encode(vae_input)
+    model_input = vae_module.reshape_data(vae_output, features)
+    if "weights" in data:
+        weights = data["weights"]
+        model_input = _expand_samples(model_input, weights)
+        # Shuffle tensor to prevent consecutive samples from being the same
+        # after expansion.
+        model_input = model_input[torch.randperm(model_input.size()[0])]
```
Comment (Ianlmgoddard) [inline src/opensynth/models/faraday/gaussian_mixture/prepare_gmm_input.py:33]:
```suggestion
        model_input = model_input[torch.randperm(model_input.size(dim=0))]
```

File: src/opensynth/models/faraday/vae_model.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                optim, T_max=self.tmax, eta_min=self.learning_rate / 5
+            )
+            return [optim], [lr_schedule]
+
+    @staticmethod
+    def reshape_data(
+        kwh_tensor: torch.Tensor, features: dict[str, torch.Tensor]
+    ) -> torch.Tensor:
+        """
+        Reshape training data to turn a concatenated training tensor.
```
Comment (Ianlmgoddard) [inline src/opensynth/models/faraday/vae_model.py:213]:
```suggestion
        Reshape training data to return a concatenated training tensor.
```

**Review by Ianlmgoddard — COMMENTED** [review]
Looks good, I think we should probably add a test for the `prepare_gmm_input` function given that it's now doing something fairly complex with the sample weighting

File: tests/models/faraday/test_faraday_model.py [hunk 6]
```diff
@@ -141,3 +145,43 @@ def test_filter_mask():
...
+    )
+    batch = TrainingData(
+        kwh=torch.rand(100, 48),
+        features={"feature_1": torch.rand(100), "feature_2": torch.rand(100)},
+        weights=torch.randint(low=1, high=5, size=(100,)),
+    )
+    encoded_batch = prepare_data_for_model(vae_module, batch)
+
+    assert encoded_batch.shape[0] == batch["weights"].sum()
+    assert encoded_batch.shape[1] == 16 + 2
```
Comment (Ianlmgoddard) [inline tests/models/faraday/test_faraday_model.py:163]: Just adding some explanation of why this is 16+2. 
```suggestion
    assert encoded_batch.shape[1] == 16 + 2 # Latent dim + number of extra features
```

Alternatively we could make it

`assert encoded_batch.shape[1] == vae_module.latent_dim + len(batch["features"].keys())`

File: tests/models/faraday/test_faraday_model.py [hunk 7]
```diff
@@ -141,3 +145,43 @@ def test_filter_mask():
...
+        class_dim=2, latent_dim=16, learning_rate=0.001, mse_weight=3
+    )
+    batch = TrainingData(
+        kwh=torch.rand(100, 48),
+        features={"feature_1": torch.rand(100), "feature_2": torch.rand(100)},
+        weights=torch.randint(low=1, high=5, size=(100,)),
+    )
+    encoded_batch = prepare_data_for_model(vae_module, batch)
+
+    assert encoded_batch.shape[0] == batch["weights"].sum()
```
Comment (Ianlmgoddard) [inline tests/models/faraday/test_faraday_model.py:162]: Would it be easy to check that the encoded batch actually has repeated rows in the way want it to, and not necessarily just got the right shape? 

So for the first row where weight !=1, check that this row is repeated `weight` times?

Comment (charlotte-avery) [reply tests/models/faraday/test_faraday_model.py]: Done, see latest commit. Note I had to abstract out the `prepare_data_for_model` function to allow for better testing.

Comment (Ianlmgoddard) [reply tests/models/faraday/test_faraday_model.py]: Nice, thank you for doing that 🙂

**Review by Ianlmgoddard — APPROVED** [review]
This looks good to me, thank you for doing this! On the order of PRs, do you think this should go in before or after #48? I'm not sure if it matters too much?
